src/classes/ADDR_Validator_TDTM.cls

```diff
@@ -37,59 +37,65 @@
 */
 public class ADDR_Validator_TDTM extends TDTM_Runnable {
-	
-	/*******************************************************************************************************
+
+    /*******************************************************************************************************
     * @description Exception raised during the address validation process.
     */
-	public class ADDR_Validator_Exception extends Exception { }
-	
-	/*******************************************************************************************************
-    * @description Indicates whether the class has been run in the current context.
-    */
-	public static Boolean alreadyRun = false;
-	
-	/*******************************************************************************************************
+    public class ADDR_Validator_Exception extends Exception { }
+
+    /*******************************************************************************************************
     * @description Verifies new or modified address records, if verification enabled in the settings.
     * @param listNew The list of Addresses from trigger new. 
     * @param listOld The list of Addresses from trigger old. 
     * @param triggerAction The trigger event (BeforeInsert, AfterInsert, etc.) that caused the class to run. 
     * @param objResult The describe for Addresses.
     * @return dmlWrapper The wrapper with the records to perform DML on. 
     */
-	public override DmlWrapper run(List<SObject> newList, List<SObject> oldList, 
-    TDTM_Runnable.Action triggerAction, Schema.DescribeSObjectResult objResult) { 
+    public override DmlWrapper run(List<SObject> newList, List<SObject> oldList,
+    TDTM_Runnable.Action triggerAction, Schema.DescribeSObjectResult objResult) {
+
+		// Look in the static list of previoysly processed Id's to determine if the record has already
+		// gone through this trigger. If so, then don't process it a second time.
+        Boolean recordsPreviouslyProcessed = false;
+        if (newList[0].Id != null && ADDR_ProcessControl.processedIdsForValidation.contains(newList[0].Id)) {
+			recordsPreviouslyProcessed = true;
+        }
         
-        if(!alreadyRun) {
-	        Addr_Verification_Settings__c settings = UTIL_CustomSettingsFacade.getAddressVerificationSettings();
-	        if ((triggerAction == TDTM_Runnable.Action.BeforeInsert || triggerAction == TDTM_Runnable.Action.BeforeUpdate) &&
-	           settings.enable_automatic_verification__c) {        	
-	        	for(SObject so : newList) {
-	        		Address__c address = (Address__c)so;
-	        		if(!system.isFuture() && !system.isBatch()) {
-	        		    address.Verification_Status__c = Label.Addr_Pending_Verification;
-	        		} else {
-	        			address.Verification_Status__c = Label.Addr_Verification_Required;
-	        		}	
-        	    }
-	        }
+        if(!ADDR_ProcessControl.checkValidationRecursionState() && !recordsPreviouslyProcessed) {
```

Comment:
Good question. I think I had left the original recursion check (changing `alreadyRun` to `checkValidationRecursionState`) and then at a later point added the the `recordsPreviouslyProcessed` verification as I got into DML > 200 tests. Testing further, it appears that the first check is not required here so I'll remove it.